src/go/k8s/pkg/resources/statefulset_test.go

```diff
@@ -0,0 +1,139 @@
+package resources_test
+
+import (
+	"context"
+	"testing"
+
+	redpandav1alpha1 "github.com/vectorizedio/redpanda/src/go/k8s/apis/redpanda/v1alpha1"
+	res "github.com/vectorizedio/redpanda/src/go/k8s/pkg/resources"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestEnsure(t *testing.T) {
+	client := fake.NewClientBuilder().Build()
+
+	err := redpandav1alpha1.AddToScheme(scheme.Scheme)
+	if err != nil {
+		t.Fatalf("Expecting no error but got %v", err)
+	}
+
+	sts := res.NewStatefulSet(client, pandaCluster(), scheme.Scheme, ctrl.Log.WithName("test"))
+
+	err = sts.Ensure(context.Background())
+	if err != nil {
+		t.Fatalf("Expecting no error but got %v", err)
+	}
+
+	actual := &v1.StatefulSet{}
+
+	err = client.Get(context.Background(), sts.Key(), actual)
+	if err != nil {
+		t.Fatalf("Expecting no error but got %v", err)
+	}
+}
+
+func TestUpdate(t *testing.T) {
+	cluster := pandaCluster()
+	sts := stsFromCluster(cluster)
+
+	stsWithDifferentReplicas := sts.DeepCopy()
+
+	var newReplicas int32 = 3333
+	stsWithDifferentReplicas.Spec.Replicas = &newReplicas
+
+	stsWithDifferentResources := sts.DeepCopy()
+	newResources := corev1.ResourceList{
+		corev1.ResourceCPU:	resource.MustParse("1111"),
+		corev1.ResourceMemory:	resource.MustParse("2222Gi"),
+	}
+	stsWithDifferentResources.Spec.Template.Spec.Containers[0].Resources.Requests = newResources
+
+	tests := []struct {
+		name		string
+		sts		*v1.StatefulSet
+		expectedUpdate	bool
+	}{
+		{"no Update", sts, false},
+		{"replicas updated", stsWithDifferentReplicas, true},
+		{"resources updated", stsWithDifferentResources, true},
+	}
+
+	for _, tt := range tests {
+		updated := res.Update(tt.sts, pandaCluster(), ctrl.Log.WithName("test"))
```

Comment:
The `Update()` function you introduced has an `OR`. So, fi any of the two returns `true` we'd pass the test. Should we be testing each of the sub-functions separately?

Reply:
yeah I guess we can also add tests for the subfunctions to have even better coverage (although I would personally say that this is fine :)).

I can add those tests once we resolve the whole FP vs mocking discussion :)

Reply:
Let's add in another change.